homeassistant/components/rainforest_raven/config_flow.py

```diff
@@ -0,0 +1,159 @@
+        async with (
+            async_timeout.timeout(5),
+            RAVEnSerialDevice(dev_path) as raven_device,
+        ):
+            await asyncio.sleep(0.05)
+            for _try in range(3, -1, -1):
+                try:
+                    # Try a few times to communicate with the device,
+                    # allowing any data already in the buffer to flush.
+                    meters = await raven_device.get_meter_list()
+                except ParseError:
+                    if not _try:
+                        raise
+                else:
+                    break
+            for meter in meters.meter_mac_ids:
+                meter_info = await raven_device.get_meter_info(meter=meter)
+                self._meter_macs[meter] = (
+                    "unknown"
+                    if meter_info.meter_type is None
+                    else str(meter_info.meter_type)
+                )
+        self._dev_path = dev_path
+
+    async def async_step_meters(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Connect to device and discover meters."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            meter_macs = []
+            for raw_mac in user_input.get(CONF_MAC, "").split(","):
+                try:
+                    mac = bytes.fromhex(raw_mac)
+                except ValueError:
+                    errors[CONF_MAC] = "invalid_mac"
+                    break
+                if len(mac) != 8:
+                    errors[CONF_MAC] = "invalid_mac"
+                    break
+                if self._meter_macs.get(mac) not in ("unknown", "electric"):
+                    errors[CONF_MAC] = "no_meters_found"
+                    break
+                if mac not in meter_macs:
+                    meter_macs.append(mac)
+            if meter_macs and not errors:
+                return self.async_create_entry(
+                    title=user_input.get(CONF_NAME, DEFAULT_NAME),
+                    data={
+                        CONF_DEVICE: self._dev_path,
+                        CONF_MAC: [m.hex() for m in meter_macs],
+                    },
+                )
+
+        discovered_macs = [m.hex() for m in self._meter_macs]
+        schema = vol.Schema(
+            {
+                vol.Required(
+                    CONF_MAC, description={"suggested_value": ",".join(discovered_macs)}
+                ): str
+            }
+        )
+        return self.async_show_form(step_id="meters", data_schema=schema, errors=errors)
+
+    async def async_step_usb(self, discovery_info: usb.UsbServiceInfo) -> FlowResult:
+        """Handle USB Discovery."""
+        device = discovery_info.device
+        dev_path = await self.hass.async_add_executor_job(usb.get_serial_by_id, device)
+        unique_id = f"{discovery_info.vid}:{discovery_info.pid}_{discovery_info.serial_number}_{discovery_info.manufacturer}_{discovery_info.description}"
```

Comment:
```suggestion
        unique_id = _generate_unique_id(discovery_info)
```

Reusing the function reduces the complexity and improves the maintainability and readability.

Reply:
I'm still not convinced this is actually more maintainable, but I took a stab at it in 9eba9f6.